src/credentials/FabricTable.cpp

```diff
@@ -1248,97 +1383,548 @@ CHIP_ERROR FabricTable::AllocatePendingOperationalKey(Optional<FabricIndex> fabr
+        RevertPendingFabricData();
+
+        // After reverting, let's fatal if possible, as this should never happen.
+        VerifyOrDie(fabricIndexToUse == *outNewFabricIndex);
+
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    mStateFlags.Set(StateFlags::kIsAddPending);
+    mStateFlags.Set(StateFlags::kIsPendingFabricDataPresent);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR FabricTable::UpdatePendingFabricCommon(FabricIndex fabricIndex, const ByteSpan & noc, const ByteSpan & icac,
+                                                  Crypto::P256Keypair * existingOpKey, bool isExistingOpKeyExternallyOwned)
+{
+    VerifyOrReturnError(mOpCertStore != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), CHIP_ERROR_INVALID_ARGUMENT);
+
+    if (existingOpKey == nullptr)
+    {
+        // If existing operational key not provided, we need to have a keystore present.
+        // It should already have an operational key pending.
+        VerifyOrReturnError(mOperationalKeystore != nullptr, CHIP_ERROR_KEY_NOT_FOUND);
+        // Make sure we have an operational key, pending or not
+        VerifyOrReturnError(mOperationalKeystore->HasOpKeypairForFabric(fabricIndex) || mOperationalKeystore->HasPendingOpKeypair(),
+                            CHIP_ERROR_KEY_NOT_FOUND);
+    }
+
+    // We should should not have a pending root when we get here, since we can't update root on update
+    VerifyOrReturnError(!mStateFlags.Has(StateFlags::kIsTrustedRootPending), CHIP_ERROR_INCORRECT_STATE);
+
+    // We should not have pending add when we get here, due to internal interlocks
+    VerifyOrReturnError(!mStateFlags.Has(StateFlags::kIsAddPending), CHIP_ERROR_INCORRECT_STATE);
+
+    // Make sure we are updating at least an existing FabricIndex
+    const auto * fabricInfo = FindFabricWithIndex(fabricIndex);
+    ReturnErrorCodeIf(fabricInfo == nullptr, CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    // Check for an existing fabric matching RCAC and FabricID. We must find a correct
+    // existing fabric that chains to same root. We assume the stored root is correct.
+    {
+        FabricIndex collidingFabricIndex = kUndefinedFabricIndex;
+        ReturnErrorOnFailure(FindExistingFabricByNocChaining(fabricIndex, noc, collidingFabricIndex));
+        ReturnErrorCodeIf(collidingFabricIndex != fabricIndex, CHIP_ERROR_INVALID_FABRIC_INDEX);
+    }
+
+    // Handle the temp insert of NOC/ICAC
+    ReturnErrorOnFailure(mOpCertStore->UpdateOpCertsForFabric(fabricIndex, noc, icac));
+    VerifyOrReturnError(SetPendingDataFabricIndex(fabricIndex), CHIP_ERROR_INCORRECT_STATE);
+
+    FabricIndex newFabricIndex = kUndefinedFabricIndex;
+    CHIP_ERROR err =
+        AddOrUpdateInner(fabricIndex, existingOpKey, isExistingOpKeyExternallyOwned, fabricInfo->GetVendorId(), &newFabricIndex);
+    if (err != CHIP_NO_ERROR)
+    {
+        // Revert partial state added on error
+        mOpCertStore->RevertPendingOpCertsExceptRoot();
```

Comment:
There is no pending root here, but I guess we're just trying to undo exactly what we did?

Do we need to undo `SetPendingDataFabricIndex`?

Reply:
I am not sure what the best cleaning-up is here. Will add a comment